extensions framework + UEFI aarch64/x86 + rpi4b + core changes/fixes

rpardini opened the pull request:
- _"if at first you don't succeed, hide all evidence that you tried"_, but done wrong.
- v2 of all patches forced pushed, fixing stuff and adressing reviews below.
- v3 force pushed. rpi4b and uefi fixes. thanks for all the screenshots and tests!
- v4 force pushed. adding uefi-arm64 and uefi-x86 **mainline kernel** build (edge builds)
- v5 is here. many fixes for kernels, firmwares, etc. **mainline kernel for rpi4b** (edge builds)
    - avoid conflict between `armbian-firmware` and `linux-firmware-raspi2` by removing the later.
    - use separate `KERNELDIR` for these 3 new kernels, until we're sure they're safe to mix.
- v6: many fixes. hopefully final version, pending test images for `edge` and `current`
- v7: small fixes.
- v8: moved to AR-1009 branch on armbian/build, for collaboration. picked Igor's commit about wifi
- _lost track of v9-v12_
  - many people contributing, thanks. branch is rebased a few times with master. 
    - *documentation* at https://github.com/armbian/documentation/pull/161
- `v13`: many fixes, disabled "reverse cross compile", full firmware for UEFI.
 
### extensions framework (née "fragments")

- this should actually change nothing at this point, just add capabilities
- the framework is implemented in lib/extensions.sh
- the "if function x exists then call x" replaced with call_extension_method()
    - +inline documentation
    - +compatibility names

### extensions framework; meta-extensions: auto-docs and sample extension gen

- 2 extensions dealing with extensibility itself
    - detect-unused-extensions: shows which extensions are enabled, but never called.
    - gen-sample-extension-docs: generates a sample empty extension & Markdown documentation for extensions

### new extension methods and features via config variables in core Armbian

- `SKIP_EXTERNAL_TOOLCHAINS=yes` - does not download or use any linaro toolchains, only build host-installed ones
- `SKIP_BOOTSPLASH=yes` - does not patch kernel for splash file
- `EXTRA_BSP_NAME=xyz` - allows for BSP variants, useful for when extensions modify the BSP
- `EXTRA_ROOTFS_MIB_SIZE=x` - add x mib's to rootfs size, for use with very small images
- `KERNEL_EXTRA_TARGETS` - what extra targets to make kernel for, default to "modules dtbs"
- `BOOTCONFIG=none` - does not build nor install u-boot; also doesn't handle bootscripts et al
- `unset KERNELSOURCE` - does not build nor install kernel, nor build initrd, nor build nor install firmware
- `ARMHF_ARCH=skip` - does not add armhf to apt/dpkg, thus pure arm64
- `SKIP_ARMBIAN_REPO=yes` - results in armbian.list.disabled in the final image
- define `APT_EXTRA_DIST_PARAMS` with apt-cacher-ng options and use it for `PACKAGE_LIST_INSTALL/REMOVE` et al
- initial support for targeting x86/amd64 UEFI and BIOS
    - some do's/don'ts for x86/amd64, like a different `UBUNTU_MIRROR` default
    - GPT/EFI(ESP) partitions (fat, `UEFISIZE=256` to enable, mount `UEFI_MOUNT_POINT=/boot/efi`, first on disk but ends
      up at `$uefipart`=15)
    - GPT/BIOS partitions (fat, `BIOSSIZE=1` to enable, second on disk but ends up at partition 14)
    - `UEFI_FS_LABEL="armbiefi"` - to set the FAT label for the EFI partition, visible in Win/Mac
    - hard-requires gdisk package host-side
- add `add_host_dependencies()` extension method; fill `EXTRA_BUILD_DEPS="pkg pkg2"` to install to host before toolchains
  download
- add `pre_prepare_partitions()` extension method, for custom partition size calculations
- add `create_partition_table()` extension method, used to do full-custom partitioning if `USE_HOOK_FOR_PARTITION=yes`
- add `post_create_partitions()` extension method, mostly for easy debugging
- add `post_write_sdcard()` extension method, where you can also set `SKIP_VERIFY=yes` to skip sdcard verification
- add `pre_install_kernel_debs()` and `post_install_kernel_debs()` extension methods.
- multiple fixes to bsp to avoid spurious errors when files are not where it expects
- v4: detect `update-initramfs` failure and abort build with useful message if it does
- v4: show useful stacktrace in `exit_with_error()` and `unmount_on_exit()`
    - if `ERROR_DEBUG_SHELL=yes`, drop into a shell before unmounting/deleting everything, so we can inspect what went
      wrong
- v4: display a message before `apt-get remove PACKAGE_LIST_BOARD_REMOVE` packages, so any errors while removing are
  easy to understand
- v4: preserve kernel .config's dates when copying

### extensions framework; refactor tool fetching/building into extensions

- a few examples of core refactoring using extensions
- `sunxi-tools` extension, enabled by 2 different `sunxi` family includes ("reuse" example)
- `marvel-tools` extension, enabled by 2 different `mvebu` family includes
- `rkbin-tools` extension, enabled by `rockship64_common` family include
- `amlogic-fip`/`c2-blobs` stuff refactored directly into `meson64_common.inc` ("single-use" example)
- removed the 'testings' fetch_from_repo completely since not used anywhere.

### .wip's for UEFI arm64 and UEFI/BIOS x86 via new GRUB extension

- v3: added `growroot`-awareness to `armbian-resize-filesystem`
    - the partition-growing part of `armbian-resize-filesystem` does not deal correctly with the UEFI layout
    - `growroot` is installed on UEFI images by default, that handles growing partition during initramfs
    - now `armbian-resize-filesystem` handles `resize2fs` only, and works.
- v4: reworked UEFI board/family/include structure:
    - use Distro's `linux-generic` kernel only for `current`
    - `edge` now builds it's own pure-mainline `5.15.y` kernel, for both x86 and arm64
        - `.config` taken from Ubuntu, probably needs tuning for EXTRAWIFI=yes et al
- v4: introduce `SKIP_KERNEL_SYMLINK=yes`, tested in `builddeb`
    - to avoid symlinking kernel; u-boot likes it, but grub and flash-kernel hates it
- v5: many fixes
- v7: more small fixes.

### .wip for the RaspberryPi 4B via new flash-kernel extension

- this does not build it's own kernel "yet", but uses default linux-raspi kernel from Ubuntu
- flash-kernel is not really a bootloader
  - it just prepares kernel et al a FAT partition for booting by the RPi4b bootloader
  - flash-kernel is standard Debian package, but has only been tested on Ubuntu releases
- it is really only known-working since Hirsute release.
  - Debian's rpi kernel is armhf only, so out of scope here, at least until we add source-built kernels.
- v3: fixed focal rootfs build. untested.
- v3: better variable names, preparing for source-built kernel.
- v5: new edge build with mainline kernel.
- v6: many fixes and some hacks for packaging and layout, also firmware (using Ubuntu's)

rpardini commented:
Moving a @The-going review, made in a separate issue, here.
-----------------------------------------------------------------------------

1) Development of the debian kernel package build script.
    Today we just copy two scripts to the source directory.
    This will be in the source package.
    
https://github.com/armbian/build/blob/1831d83acc9ef651f4f4d38eaccdbc8d2357c2b0/packages/armbian/builddeb#L291-L303

   How will this global variable be initialized when the kernel is built natively, without the armbian build system from the source package?

rpardini commented on packages/armbian/builddeb:
```diff
@@ -289,15 +289,17 @@ EOF
-touch /boot/.next
-exit 0
-EOT
+if [ "${SKIP_KERNEL_SYMLINK}" != "yes" ]; then
+	##
+	## Create sym link to kernel image
+	##
+	sed -e "s/exit 0//g" -i $tmpdir/DEBIAN/postinst
+	cat >> $tmpdir/DEBIAN/postinst <<- EOT
+	ln -sf $(basename $installed_image_path) /boot/$image_name 2> /dev/null || mv /$installed_image_path /boot/$image_name
+	touch /boot/.next
+	exit 0
+	EOT
+fi
```
Original review by @The-going:
---------------------------------------------------------------
Development of the debian kernel package build script.
    Today we just copy two scripts to the source directory.
    This will be in the source package.
   How will this global variable be initialized when the kernel is built natively, without the armbian build system from the source package?

rpardini replied:
Hmm. I had _not_ considered this would be built outside the build system, but is indeed valid question.
Anyway, the same question could be asked of ALL the variables passed in https://github.com/armbian/build/blob/1831d83acc9ef651f4f4d38eaccdbc8d2357c2b0/lib/compilation.sh#L513-L530

I would much prefer if we used the standard builddeb from upstream kernel, since that does not do that symlink.

rpardini replied:
Reply from @The-going:
----------------------------------------------------------------------------------------------------------------

Good health, Ricardo.
> Anyway, the same question could be asked of ALL the variables passed in

That's right, you're right. It is for this reason that the discussion has a general character.

> I would much prefer if we used the standard builddeb from upstream kernel, since that does not do that symlink.

[Standard scripts](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/scripts/package) from the kernel sources are very good and working examples\templates,
but this is not enough for industrial release.
Armbian has improved these scripts so that they do a little more.
My question is not how you handled the exception today and removed the creation of a symbolic link, but how we plan to do it in the future. I want to hear your opinion as an author.
I hope I wasn't too intrusive.

rpardini replied:
Not intrusive at all! Very much appreciated. @The-going 
I think it might be interesting to handle the packaging maybe as an extension itslef, so different packaging needs might have different solutions. For example, for UEFI, we could maybe just use the standard packaging!  I will propose solution in a separate commit for review, in this same PR/branch.

The-going replied:
We should try to avoid using standard packaging altogether.
This is my inner conviction.

rpardini replied:
Addressed by removing the hack, and fixing the missing `$image_name` that was the original problem.
Thanks @The-going  for driving this investigation and de-hacking !

rpardini commented:
So: #3282 died and this is it's heaven. Let's just continue as if nothing happened?
Thanks.

rpardini reviewed (commented):
From @The-going:
-----------------
Good health, Ricardo.
> Anyway, the same question could be asked of ALL the variables passed in

That's right, you're right. It is for this reason that the discussion has a general character.

> I would much prefer if we used the standard builddeb from upstream kernel, since that does not do that symlink.

[Standard scripts](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/scripts/package) from the kernel sources are very good and working examples\templates,
but this is not enough for industrial release.
Armbian has improved these scripts so that they do a little more.
My question is not how you handled the exception today and removed the creation of a symbolic link, but how we plan to do it in the future. I want to hear your opinion as an author.
I hope I wasn't too intrusive.

The-going commented:
Good. But not very fast. I translate for a long time.

heisath commented:
I am not much into the build script itself, but tested the PR on mvebu. Everything works and it is nice to have the good debugging visibility of extension calls.

Just one thing, I am not sure if normal mvebu family (without mvebu64) does even need the marvell tool extension. Would need to check that, but not urgent :)

rpardini commented:
> Just one thing, I am not sure if normal mvebu family (without mvebu64) does even need the marvell tool extension. Would need to check that, but not urgent :)

Indeed. I did not build mvebu or mvebu64 myself, so great that you did, thanks! Just remove the `enable_extension` call from non-64 mvebu, and if it still builds, we can remove it.

The-going commented:
I am currently studying your code and find it very mature. And this is very good.
I have certain plans to improve the quality of the created debian packages. Scripts for building packages for the kernel is a very big topic. We can improve what we have or we can take that as a basis. what already exists in debian\ubuntu distributions. At least their methodology \ algorithm.

We need to decide and understand how far the participants of Armbian are ready to advance in this matter.

rpardini commented:
> Scripts for building packages for the kernel is a very big topic. We can improve what we have or we can take that as a basis. what already exists in debian\ubuntu distributions. At least their methodology \ algorithm.

Ubuntu packaging is at https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux-meta/+git/impish?h=master but very complex since they handle everything in there, dkms, firmware, nvidia driver crazyness, etc.

For now I will try to make minimal changes to what we already have. Maybe just write the values of the variables to a file and source it from the packaging? That way the source package includes them, and can be built outside of the build system.
 
> We need to decide and understand how far the participants of Armbian are ready to advance in this matter.

Yes. It's very important. In this PR, [this commit](https://github.com/armbian/build/pull/3300/commits/99b9da6357cdfb2a4b0ba6ecefdc9f04c03e1f16) has examples of moving some simple stuff from core into extensions. Same can/should be done by others in many other parts of the system, there's all kinds of small or big stuff to work on.

Tonymac32 commented:
@lanefu said I should come provide input.  😆

I like it, and think we should decide if the existing PR is what we want to work from, merge it, then debate future actions.

rpardini commented:
> @lanefu said I should come provide input. 😆

Thanks for coming sir, the party is just starting... 🎉  some suggestions for fun to be had:

-  Checkout the branch and build a few boards with it when/if you have the time...
- Then add `ENABLE_EXTENSIONS=gen-sample-extension-docs` and build again, check out the generated `userpatches/extensions/sample-extension.sh`, check out `output/debug/hooks.auto.docs.md`, etc...
- **Bonus!!** if you build _again_ with `ENABLE_EXTENSIONS=sample-extension` and then write a simple extension based on it. 

(Rockship, for one, has not been tried with this that I know of, that would be very valuable input indeed... and, it should feel good to build Rockship without downloading FIP trees for Amlogic, finally.)

lanefu reviewed (approved):
Let's get this merged!   I know it's a big change, but it'd ls nicely documented and frankly is going to play a big part in our future success

lanefu commented:
I have certain plans to improve the quality of the created debian packages. Scripts for building packages for the kernel is a very big topic. We can improve what we have or we can take that as a basis. what already exists in debian\ubuntu distributions. At least their methodology \ algorithm.
> 
> We need to decide and understand how far the participants of Armbian are ready to advance in this matter.

This should be a seperate conversation.  We can discuss after this is merged

lanefu commented:
@rpardini  are you ready to remove draft status?

150balbes commented:
IMHO. Until it is thoroughly and fully tested on different devices and build modes, with different variables and kernels, it should not go to the master (it should remain a separate branch). It is very easy to add new things, but when there are errors and problems, it is very difficult to roll back the changes.

150balbes reviewed (approved):
IMHO. Until it is thoroughly and fully tested on different devices and build modes, with different variables and kernels, it should not go to the master (it should remain a separate branch). It is very easy to add new things, but when there are errors and problems, it is very difficult to roll back the changes.

igorpecovnik commented:
> Until it is thoroughly and fully tested on different devices and build modes

We are testing this for a while ... Today I also made rootfs caches from this branch:
https://github.com/armbian/build/actions/runs/1520015520
and it looks fine, except x86 builds. Which I am about to correct during a day. It is important that this merge doesn't break the existing code - possible bugs in x86 or rpi are actually not critical for merge.

rpardini commented:
> Until it is thoroughly and fully tested on different devices and build modes, with different variables and kernels

Apart from Igor's efforts, I've been building my variants (all new boards, plus some old, plus most of meson64) with an all-green run here https://github.com/rpardini/armbian-release/actions/runs/1518733458

The-going commented:
The SKIP_KERNEL_SYMLINK key is valid when the kernel configuration contains UEFI support as CONFIG_EFI=y, or is there another case without UEFI support that you don't need to create a symbolic link?

rpardini commented:
> The SKIP_KERNEL_SYMLINK key is valid when the kernel configuration contains UEFI support as CONFIG_EFI=y, or is there another case without UEFI support that you don't need to create a symbolic link?

Actually the only reason I removed the symlink is because it fails on UEFI builds, it produces a circular symlink to itself.
I did not look much into it, to say the truth, until I found and removed the hidden `make clean` in the headers packaging. That was slowing me down too much, every packaging change needed a full kernel rebuild and that tooks hours for me. I could look into this during the weekend.

Anyway I like the idea of testing the kernel config for EFI and using that instead of external variable to decide. Thanks.

The-going commented:
I will handle this exception correctly. You don't have to do this.

I will ask you to do an additional check of the current kernel configuration of the CONFIG_EFI=y parameter before the UEFI-enabled assembly steps begin.

The-going commented:
In fact, we need to check whether or not the user has enabled UEFI support in the kernel configuration when he builds an image with GRUB\UEFI support.

rpardini commented:
Last commit allows building amd64 kernels on aarch64 hosts. _For the brave only. (Kernel building is ok, but building amd64 rootfs on aarch64 via qemu is painfully slow, much more than the opposite in my experience)._

rpardini commented:
> (and as far as I know, has never been involved in maintenance and responses) it is a pleasant and easy job

I won't comment on the rest, but that is very unfair, sir. I've been involved in maintenance and fixing of bugs of meson64 at least for the last year, if not more: https://github.com/armbian/build/commits?author=rpardini

On maintenance: if you notice, `current` uses the Distro's kernel and firmware. That requires almost zero maintenance, since it just works exactly like Ubuntu does. Worst-case scenario, we drop the `edge` version with kernel building and patches.

150balbes commented:
> I won't comment on the rest, but that is very unfair, sir. I've been involved in maintenance and fixing of bugs of meson64 at least for the last year, if not more

Don't be offended, I didn't mean to offend you. I expressed how I see the situation. :)

lanefu commented:
> OFFTOPIK I don't pretend to be the truth, but I have big concerns. :)
> 
> we have not yet had time to deal with the current problems of overload from inadequate users, as you add a new problem. There will be a huge flurry of appeals, squeals and demands from users, x86 and RPi, who are used to getting everything for free and at their first request. A huge number of people and large financial flows work to create and support Debian\Ubuntu for x86 and rpi, but this is not the case for Armbian. You are opening pandora's box. :) Have you thought well ? pardini creates new code (and as far as I know, has never been involved in maintenance and responses) it is a pleasant and easy job, it is many times more difficult to support and maintain what has been created. Who will do this ?

It won't be supported it will be CSC until a maintainer volunteers.

We've been communicating that we are changing how support works and what we provide.

https://www.armbian.com/newsflash/armbian-needs-your-help/

https://docs.armbian.com/User-Guide_Board-Support-Rules/

igorpecovnik commented on config/sources/families/include/uefi_common.inc:
```diff
@@ -0,0 +1,23 @@
+	export KERNELPATCHDIR="uefi-${LINUXFAMILY}-${BRANCH}" # Might be empty.
+	export KERNELDIR="linux-uefi-${LINUXFAMILY}"          # Avoid sharing a source tree with others, until we know it's safe.
+	function add_host_dependencies__libelf_and_libfdt() { # Conditional extension: when building the UEFI kernels, make sure some host-side libs are available
+		export EXTRA_BUILD_DEPS="${EXTRA_BUILD_DEPS} libfdt-dev libelf-dev"
```
I would propose to keep host dependencies as is. Always have them all - and when new one is needed, it should also be added to the Dockerfile.

rpardini replied:
Yes, in this case it's small packages. It's something big like the x86 toolchain for aarch64 I'd keep it.

igorpecovnik replied:
Compilers are different, yeah - they are anyway tied to some specific source and should be downloaded if / when needed. Current implementation downloads them all since we didn't bother with this at the time. But extensions can help to deal with this in a better way. That's cool. Just not fragmenting general host dependencies more.

rpardini replied:
Agreed. Done.

The-going commented:
> I did not look much into it, to say the truth, until I found and removed the hidden `make clean` in the headers packaging. That was slowing me down too much, every packaging change needed a full kernel rebuild and that tooks hours for me.

Did I understand correctly? Do we need to handle the exception and not clean up the script directory for the header package if we are building for the x86\amd64 architecture? Or are we just not collecting this package?

rpardini commented:
> > I did not look much into it, to say the truth, until I found and removed the hidden `make clean` in the headers packaging. That was slowing me down too much, every packaging change needed a full kernel rebuild and that tooks hours for me.
> 
> Did I understand correctly? Do we need to handle the exception and not clean up the script directory for the header package if we are building for the x86\amd64 architecture? Or are we just not collecting this package?

Please take a look. This is in master armbian/build, also in this branch. Specially line 341, during the header packaging, it does a "make clean", throwing away all the work of 1 hour compilation time. It cleans the build tree _after building it_, not _before starting a new build_, and there's no way to control it. So every build is a rebuild, even if `CLEAN_LEVEL=debs`. Why?
  https://github.com/armbian/build/blob/39263b5e76e3209d150ab8ddd558ff4ffd92360f/packages/armbian/builddeb#L338-L343

The-going commented:
I'm surprised! I did a few kernel builds today.This is not happening.
This command should only clear one script folder. For which version of the kernel is this happening for you?

rpardini commented:
> I'm surprised! I did a few kernel builds today.This is not happening. This command should only clear one script folder. For which version of the kernel is this happening for you?

Yes, people with 32+ core machines hardly notice. For me is suffering. I'm building 5.15.5. The build tree goes from 4Gb+ to 1.5Gb after that clean.

The-going commented:
When I have, the compiler cache is clean, the kernel assembly takes 25-30 minutes. All the next 3-5 minutes.

rpardini commented:
Rebased master. No other changes.

rpardini commented:
> When I have, the compiler cache is clean, the kernel assembly takes 25-30 minutes. All the next 3-5 minutes.

Yeah, for me is 1h+ and 15 minutes. If I comment out that clean, second build is few minutes.

The-going commented:
Are you ready to deal with the `C' compiler cache problem?

rpardini commented:
> Are you ready to deal with the `C' compiler cache problem?

I don't have such a problem. Actually, in my own fork, I've just commented out the `make M=scripts clean` and everything works exactly how I want it. I just thought other developers might benefit from faster rebuilds too.

rpardini commented:
Wrote some documentation. 🎉  It's very early prototype, still being converted from old _fragments_ stuff.

Intention is to get it included in docs, so a separate repo, separate PR: https://github.com/armbian/documentation/pull/161

The-going commented:
> I don't have such a problem. Actually, in my own fork, I've just commented out the `make M=scripts clean` and everything works exactly how I want it. I just thought other developers might benefit from faster rebuilds too.

If I understand correctly, are you using a builddeb file, the one that is in the build system, but have you commented out this line?
Then please tell me for which assembly line you use it as:
```
[ o.k. ] Repeat Build Options [ ./compile.sh  BOARD=virtual-qemu BRANCH=current RELEASE=focal ...
```

rpardini commented:
> I'm starting to check 5.15
> I found a problem. This problem is in two places. No such behavior has been observed for kernel 5.13 (this needs to be checked additionally). And when there is a native compilation, we cannot clear the scripts folder. This results in a complete cleanup of the build directory.

So I'm not crazy? Bless you.
Maybe, it's not _that_ make clean, but some other thing somewhere? I've looked really hard and didn't find anything else. I've tried `IGNORE_UPDATES=yes` and that made no difference, so it was not `git clean`.

The-going commented:
I know. This script is intended only for cross-compilation and takes into account only these features as:
x86_64 -> arm\arm64.
But added x86_64 ->x86_64,x86; ...

The-going commented:
You built for x86_64. Have you checked the package with the headers?

The-going commented:
`arch/x86/Makefile:142: CONFIG_X86_X32 enabled but no binutils support`
I install:
`binutils-x86-64-linux-gnux32`
but there is no effect

rpardini commented:
@The-going I found the problem I was facing originally, please take a look at https://github.com/armbian/build/pull/3300/commits/9138514e4d59507338f8bb19988d09f7c5d73a90
I've removed the SKIP_KERNEL_SYMLINK hack, returned to original. Now UEFI (x86 and aarch64) can use the same packaging.

rpardini commented:
> `arch/x86/Makefile:142: CONFIG_X86_X32 enabled but no binutils support` I install: `binutils-x86-64-linux-gnux32` but there is no effect
Hmm are you cross compiling from aarch64 to x86-64? 
My arm64 build host has `binutils-x86-64-linux-gnu` (not x32)

The-going commented:
> @The-going I found the problem I was facing originally, please take a look at [9138514](https://github.com/armbian/build/commit/9138514e4d59507338f8bb19988d09f7c5d73a90) I've removed the SKIP_KERNEL_SYMLINK hack, returned to original. Now UEFI (x86 and aarch64) can use the same packaging, since grub only looks at the main name.

You did everything right, but
it doesn't solve the problem in full. Look in the header package.

rpardini commented:
> >  I've removed the SKIP_KERNEL_SYMLINK hack, returned to original. 
> You did everything right, but it doesn't solve the problem in full. Look in the header package.

Hmm, it looks like it's working. If it isn't, it's a separate problem. 

![Screenshot 2021-12-01 at 22 15 14](https://user-images.githubusercontent.com/639959/144315433-a92d433a-ef65-40b4-8b19-6dd2e367e176.png)

We should split the cleaning problem to a separate issue, since it also affects master.
And headers for x86 might be handled later too.
What do you think?

The-going commented:
```
dpkg-architecture -L | grep x86
or
dpkg-architecture -L | grep arm
```
https://github.com/armbian/build/blob/1cf80ec462e6ef15e8bab0e983e24bf115a7752e/lib/compilation.sh#L414-L415

I've almost finished everything. There are two more tests tomorrow. That's enough for today.

rpardini commented:
> https://github.com/armbian/build/blob/1cf80ec462e6ef15e8bab0e983e24bf115a7752e/lib/compilation.sh#L414-L415

Yes, but please, `git pull --rebase`, that has been addressed yesterday by https://github.com/armbian/build/pull/3300/commits/0c5ee20bb1b33a133e6e359476082d43d5ad457c (x86 should exclusively be built by system toolchains, which are in the default PATH).

> I've almost finished everything. There are two more tests tomorrow. That's enough for today.

Thank you @The-going, sir, for the review. Much appreciated.

The-going commented:
> dpkg-architecture -L | grep x86
> or
> dpkg-architecture -L | grep arm

Before compilation begins, we check the contents of the ARCHITECTURE variable for compliance with the requirements of debian, if it does not, then we end with an error.

Then we check the architecture of the machine we are on and compare it with $ARCHITECTURE. If this is a native compilation, then just initialize the empty variables ARCH, KERNEL_COMPILER, or specify values corresponding to the build environment of the current machine.

The build system `debuild` and all its subordinates as 'dpkg-*' uses its own runtime environment and a large set of its own variables. When build is in its own environment, all the necessary variables are calculated automatically.

rpardini commented:
> Please remove the reverse cross compilation from this PR. i.e. aarch64 -> x86_64. 

@The-going I'm not sure I follow. 

It should be working perfectly, if not, please point me to how and where it fails for you.

All these images (https://github.com/rpardini/armbian-release/releases/tag/20211130b) were built on an Ampere N1 aarch64 machine, including the x86_64 images, and seem to boot and run just as good as others I've built on an x86_64 machine. So cross-compile and native compile should work equally well.

Please explain a bit more about this.

lanefu commented:
@The-going  It is diffcult as it'sa  very technical conversation.  Try explaining in Russian and we'll see if @iav  can help instead of the Robot :)

@iav Would you mind trying to help understand what The-Going's worries are?

igorpecovnik commented:
> aarch64 -> x86_64. This requires additional effort 

Yes, but it's probably worth going that direction. 

Our automation and build framework is CC since the day one and we try to move on compiling natively on aarch64. We are not at 100% there yet, but close. Our CI engine is already building on both, so in case aarch64 server happens to get this job, he will build it, but slower. If this proves to just bring troubles, we can remove in the future.

iav commented:
I will glad to help, but now I not understand the root of worries.
But in case the language can help, let's try:

@The-going, в чём конкретно вы видите настолько серьёзную проблему, что считаете предпочтительным выкинуть кусок сборки aarch64→x64? Только то, что он медлено работает на aarch64, или что-то ещё, или что-то совсем другое? 
Как человек, в основном использующий для сборки платформу aarch64, считаю, что «медленно — это гораздо лучше, чем никак».

rpardini commented:
I built all my images again last night, using exclusively **aarch64 build machines**, so cross compilation in full effect for the x86 builds.
These include all the commits fixes up to this point.

`v10` AR-1009 based `edge` (**mainline kernel**) test images are up! Using `5.15.6` kernel.

- `RaspberryPi  4b`: 
  - [Armbian_20211202x-rpardini_Rpi4b_impish_edge_5.15.6.img.xz](https://github.com/rpardini/armbian-release/releases/download/20211202x/Armbian_20211202x-rpardini_Rpi4b_impish_edge_5.15.6.img.xz)
- `UEFI x86 (Desktops, Intel Macs, x86 SBCs, etc)`: 
  - [Armbian_20211202x-rpardini_Uefi-x86_impish_edge_5.15.6.img.xz](https://github.com/rpardini/armbian-release/releases/download/20211202x/Armbian_20211202x-rpardini_Uefi-x86_impish_edge_5.15.6.img.xz)
- `x86 VMWare Player/Fusion/Workstation`: 
  - [Armbian_20211202x-rpardini_Uefi-x86_impish_edge_5.15.6.vmware.uefi.zip.xz](https://github.com/rpardini/armbian-release/releases/download/20211202x/Armbian_20211202x-rpardini_Uefi-x86_impish_edge_5.15.6.vmware.uefi.zip.xz)
- `UEFI aarch64 (SolidRun Honeycomb, Pythium, etc)`: 
  - [Armbian_20211202x-rpardini_Uefi-arm64_impish_edge_5.15.6.img.xz](https://github.com/rpardini/armbian-release/releases/download/20211202x/Armbian_20211202x-rpardini_Uefi-arm64_impish_edge_5.15.6.img.xz)


(many more builds here, also `current` with Distro's kernel and clouds:) [20211202x](https://github.com/rpardini/armbian-release/releases/tag/20211202x)

The-going commented:
@iav , @150balbes Ru:
Сегодня мы собираем пакеты с помощью кросс компиляции на x86_64 машине для машин aarch64 и arm*. И при этом используем хитромудрый\нетрадиционный приём когда создаём пакет с заголовками. Мы очищаем папку скриптов,
https://github.com/armbian/build/blob/aed2717909f2386c133c5cd481a86a5e80782f82/packages/armbian/builddeb#L339-L341
то есть тот код который скомпилирован для x86_64 машины и который выполняется в процессе создания пакета для arm машины. Но чтобы пакет с заголовками получился полноценным нам приходится в последствии накладывать\применять этот патч.

https://github.com/armbian/build/blob/aed2717909f2386c133c5cd481a86a5e80782f82/packages/armbian/builddeb#L135-L138

И компилировать их снова уже находясь на целевой машине. Для этого в пакете заголовков существует скрипт
https://github.com/armbian/build/blob/aed2717909f2386c133c5cd481a86a5e80782f82/packages/armbian/builddeb#L75-L84
Для родной нативной компиляции, эти действия необходимо исключить. И этого достаточно. Но скрипт должен понимать, что от него требуется. Для этого необходимо корректно обработать исключения.

Теперь добавим сюда перекрёстную компиляцию родственных машин
aarch64 -> armhf,
x86_64 -> i386
Очевидно, что логика усложняется.
И напоследок, обратная компиляция.
Нам придётся добавить патчи для x86_64 машины, очищать папку скриптов ..., когда мы собираем пакет заголовков на машине aarch64. Всё очень усложняется.

The-going commented:
> test images are up!

Image yes but not header packages!

150balbes commented:
> using exclusively **aarch64 build machines**

Write the commands that you use to build x86, I want to test the build speed on my two device variants ARM (on a device with maximum performance and on a device with minimal parameters).

The-going commented:
> Как человек, в основном использующий для сборки платформу aarch64, считаю, что «медленно — это гораздо лучше, чем никак».

Я придерживаюсь традиции собирать пакет для ахитектуры на сомой архитектуре. Т.е. родная компиляция. Это исключит много подводных камней.

rpardini commented:
> > test images are up!
> 
> Image yes but not header packages!

The images _contain_ the header package. If you boot them up, you will find that `/usr/src` contains the headers package (`.deb`), ready to be installed. This is a trick made possible by a user extension, but lives only in my fork/branch.

The-going commented:
> This is a trick made possible by a user extension, but lives only in my fork/branch

Good. And now build the kernel module using this package.

150balbes commented:
> Сегодня мы собираем пакеты с помощью кросс компиляции на x86_64 машине для машин aarch64 и arm*.

не совсем так  :)
я собираю часть образов непосредственно на железках ARM (кстати, я еще собираю версии альтлинукс и только на самих железках ARM) и нативная сборка работает не плохо.

> Для этого необходимо корректно обработать исключения.

Это легко реализуется, посмотрите, как в скриптах сборки отрабатывается проверка где запущена сборка и выбирается соотвествующий вариант действий. Нужно "просто" найти время (это самое друдное, найти время) и подправить скрипты сборки. Либо вообще использовать два разных и вызывать их по определению, на какой архитектуре запущена сборка.

---------------------------------

> Today we are building packages using cross compilation on an x86_64 machine for aarch64 and arm * machines.

Not certainly in that way  :)
I collect some of the images directly on the ARM hardware (by the way, I also build the altlinux versions and only on the ARM hardware themselves) and the native assembly works not badly.

> To do this, you must correctly handle the exceptions.

This is easy to implement, see how in the build scripts the check is processed where the build is started and the appropriate option is selected. You just need to find the time (this is the most difficult thing, to find the time) and tweak the build scripts. Or, in general, use two different ones and call them by definition on which architecture the assembly is running.

The-going commented:
My suggestion. We need to adopt the infrastructure proposed by the author, what has already been implemented in this PR. I will add logic for the correct assembly of headers and then we will test each line for a long time and correct possible roughness. No need to add new functionality before that.

rpardini commented:
> Good. And now build the kernel module using this package.



> Good. And now build the kernel module using this package.

Да! Спасибо, сэр, я наконец понял, о чем вы. Я буду работать над этим и придумываю исправление как можно скорее.

Также спасибо всем, кто помог мне понять.

The-going commented:
@rpardini Just fix\stop this PR. Let me make corrections. And we will continue after the merger.

rpardini commented:
> @rpardini Just fix\stop this PR. Let me make corrections. And we will continue after the merger.

Ok, go ahead sir. I will not push commits until you give me the green light.

The-going commented:
> Это легко реализуется, посмотрите, ...
Всё дело в стратегии развития. Один и тот же результат может быть достигнут разными путями.
Обсуждать это пямо здесь, в той точке где проблема должна быть решена, нехотят.
И мне самостоятельно надо выбрать этот путь. Тогда просто зафиксируем это как есть, немного подправим и обединим.
Но пакеты в любом случае должны оставаться рабочими.

rpardini commented:
> > using exclusively **aarch64 build machines**
> 
> Write the commands that you use to build x86, I want to test the build speed on my two device variants ARM (on a device with maximum performance and on a device with minimal parameters).

@150balbes You can simply build `BOARD=uefi-x86 BRANCH=edge`  -- if on aarch64 host it should depend on the correct toolchain package host-side, and build the kernel correctly with it. It's not much slower than native compile. 

What is very slow is the deboostrap/rootfs build due to qemu emulation (and being mostly single thread, but that's not specific to cross build).

rpardini commented:
> Всё дело в стратегии развития. Один и тот же результат может быть достигнут разными путями.
> Обсуждать это пямо здесь, в той точке где проблема должна быть решена, нехотят.
> И мне самостоятельно надо выбрать этот путь. Тогда просто зафиксируем это как есть, немного подправим и обединим.
> Но пакеты в любом случае должны оставаться рабочими.

The problem @The-going found is valid (again! thanks sir), when the `kernel-headers-` package is cross built for x86 target on aarch64 host, the resulting .deb when installed on a x86 host will currently *not* work. The same does not happen if kernel is built on x86. 

> My suggestion. We need to adopt the infrastructure proposed by the author, what has already been implemented in this PR.

I agree that adding "one more hack" to fix that very specific problem might not be the best way forward. Especially when I'm proposing an extension framework in the same PR. 

Also with a complete, well thought solution we might be able to better reuse previous build's output for rebuilds by removing the  workaround currently in use, if I understand correctly. That would greatly speed up the development cycle of the packaging itself, and when testing patches, and thus is good for everyone.

Let's see what we can come up with in terms of solutions, thanks @iav @150balbes @The-going for helping me understand this and @lanefu @igorpecovnik for the support.

The-going commented:
All done in  #3313

rpardini commented:
Merged @The-going's PR. Thank you!

The-going commented:
I did  build for (virtual-qemu:current) on a local machine. And it was successful. But I noticed that cleaning before assembling the header package for the 5.10 kernel differs from cleaning for 5.15. In the first case, only the script directory is cleaned, in the second it looks like a complete cleaning. We will have to check the workability of the header package for the 5.15 kernel.

rpardini commented:
Ok, I will do a full round of builds and tests. 
@The-going  please check https://github.com/armbian/build/pull/3300/commits/080c9381153aaedf5246960102b6e76b45df4220

- packaging:
  - there is _no need_ anymore for the symlink hack, CONFIG_EFI or no. But check is great, see below
  - it's not `amd64` that has no DTB's, it's all UEFI, thus: `is_enabled CONFIG_EFI`, thanks!
- Explicitly disallow "reverse cross compile" in amd64.conf.

The-going commented:
Logic is not ideal and if you see errors or have ideas, just call. I'm in touch.

rpardini commented:
Thank you @The-going ! I will do a full round of builds and test images soon.

@lanefu has graciously provided Armbian with x86 builder on lanecloud for this PR so I don't need to do _reverse cross compilation_ anymore.

rpardini commented:
I will also rebase with master before the new round of builds.
In the process the merge commit will be eliminated, thus producing a linear history

rpardini commented:
Turns out I was naive, many families already had `CONFIG_EFI=y` in their .config. Don't ask me why.
I will revert the logic that disables dtb package generation back to @The-going's original (`if ! amd64`)

rpardini commented:
Fixes done.
Also added `-full` firmware for UEFI boards, and made it "really" full by not removing anything anymore.

rpardini commented:
Pushed another fix, firmware related.
Current version deemed `v12`.

rpardini commented:
Another rebase to master to pickup critical bugfix there.
Dubbed `v13`. Images coming soon.

rpardini commented:
`v13` AR-1009 based `edge` (**mainline kernel**) test images are up! Using `5.15.6` kernel, "Full" (amdgpu,etc) firmware for UEFI, and many fixes. Thanks to all who contributed.

- `RaspberryPi  4b`: 
  - [Armbian_20211205c-rpardini_Rpi4b_impish_edge_5.15.6.img.xz](https://github.com/rpardini/armbian-release/releases/download/20211205c/Armbian_20211205c-rpardini_Rpi4b_impish_edge_5.15.6.img.xz)
- `UEFI x86 (Desktops, Intel Macs, x86 SBCs, etc)`: 
  - [Armbian_20211205c-rpardini_Uefi-x86_impish_edge_5.15.6.img.xz](https://github.com/rpardini/armbian-release/releases/download/20211205c/Armbian_20211205c-rpardini_Uefi-x86_impish_edge_5.15.6.img.xz)
- `x86 VMWare Player/Fusion/Workstation`: 
  - [Armbian_20211205c-rpardini_Uefi-x86_impish_edge_5.15.6.vmware.uefi.zip.xz](https://github.com/rpardini/armbian-release/releases/download/20211205c/Armbian_20211205c-rpardini_Uefi-x86_impish_edge_5.15.6.vmware.uefi.zip.xz)
- `UEFI aarch64 (SolidRun Honeycomb, Pythium, etc)`: 
  - [Armbian_20211205c-rpardini_Uefi-arm64_impish_edge_5.15.6.img.xz](https://github.com/rpardini/armbian-release/releases/download/20211205c/Armbian_20211205c-rpardini_Uefi-arm64_impish_edge_5.15.6.img.xz)


[All images at 20211205c](https://github.com/rpardini/armbian-release/releases/tag/20211205c) 

I'll do some testing tomorrow and unless something ugly comes up I'll put this PR for final review.

rpardini commented:
`v13` AR-1009 based `edge` (**mainline kernel**) test images are up! 

Using `5.15.6` kernel, "Full" (amdgpu,etc) firmware for UEFI, and many fixes. Thanks to all who contributed.

- `RaspberryPi  4b`:
    - [Armbian_20211205d-rpardini_Rpi4b_impish_edge_5.15.6.img.xz](https://github.com/rpardini/armbian-release/releases/download/20211205d/Armbian_20211205d-rpardini_Rpi4b_impish_edge_5.15.6.img.xz)
- `UEFI x86 (Desktops, Intel Macs, x86 SBCs, etc)`:
    - [Armbian_20211205d-rpardini_Uefi-x86_impish_edge_5.15.6.img.xz](https://github.com/rpardini/armbian-release/releases/download/20211205d/Armbian_20211205d-rpardini_Uefi-x86_impish_edge_5.15.6.img.xz)
- `x86 VMWare Player/Fusion/Workstation`:
    - [Armbian_20211205d-rpardini_Uefi-x86_impish_edge_5.15.6.vmware.zip.xz](https://github.com/rpardini/armbian-release/releases/download/20211205d/Armbian_20211205d-rpardini_Uefi-x86_impish_edge_5.15.6.vmware.zip.xz)
- `UEFI aarch64 (SolidRun Honeycomb, Pythium, etc)`:
    - [Armbian_20211205d-rpardini_Uefi-arm64_impish_edge_5.15.6.img.xz](https://github.com/rpardini/armbian-release/releases/download/20211205d/Armbian_20211205d-rpardini_Uefi-arm64_impish_edge_5.15.6.img.xz)


[All images, incl `current` and cloud at 20211205d](https://github.com/rpardini/armbian-release/releases/tag/20211205d)

Tests are looking decent, and unless something ugly comes up I'll put this PR for final review soon.

igorpecovnik commented:
Bug (merged with master as well) https://pastebin.com/3k9Ld7Mg

rpardini commented:
Rebased from master again.
Added `extensions: introduce cleanup_extension_manager() called by build-all-ng's unset_all()` to fix problems with multiple board building, this cleanup's the extension manager and the extension's functions so it can run again.

rpardini commented:
Many issues identified with `build_all_ng`, PR set on draft again for some rework.

rpardini commented:
after a long debugging session with @igorpecovnik we found a nasty bug when `build-all-ng` is used in combination with extensions. simple fix implemented. PR will be up for final review.

igorpecovnik reviewed (approved):
Tested many times, we have CI operational ... lets merge.

150balbes commented:
After adding this PR, a serious error appeared. The DTB package is put into the host system and interferes with the GRUB settings of the host system. This grossly disrupts the operation of the host system and can lead to a complete blocking of the startup, with all the ensuing consequences.

igorpecovnik commented:
On arm build host, intel or both? I understand its not good, but why would this compromise grub in any way?

150balbes commented:
Host x86

igorpecovnik commented:
My main dev host is clean, two of the random runners I have checked as well. One x86, one arm64. Perhaps this leak was induced in some other situation or it happens under certain conditions? Which we need to find. In any case, let's keep an eye on.

rpardini commented:
> After adding this PR, a serious error appeared. The DTB package is put into the host system (dir /boot host) and interferes with the GRUB settings of the host system. This grossly disrupts the operation of the host system and can lead to a complete blocking of the startup, with all the ensuing consequences.

Very interesting. Will investigate deeply. The funny parts
- the-Going made the packaging completely skip the DTB package for x86, so it shouldn't ever exist to begin with
- such a non-existing DTB package was installed _on the builder host?_ - extremely curious! and dangerous. 
- dunno how it happened, if it doesn't exist to begin with, but indeed, if such a DTB package is installed, grub does go insane and causes problems (i've discovered this on _target_ images during development, never on host though)

Anyway I will investigate, thanks for the report.

150balbes commented:
I could be wrong. But perhaps there should be ready-made packages with DTB files in the out directory (for example, from previous builds of images for aarch64), then these DTB packages are used when updating packages in the host system (I noticed that this happens when installing new packages in the host system, for example, the composition of dependencies for the build environment has changed).

rpardini commented:
I did introduce a hook for adding to `hostdeps`, but can not imagine how this could cause a local .deb to be installed host-side. I will try re-enabling the DTB package to see if I can reproduce it somehow. It could also be that the same old version that produced the DTB package also installed it, but I can't see how.
Come to think of it, I think I had a similar situation where the host's initrd (not dtb) was somehow destroyed (that also drives host-side grub crazy, trying to boot directly from rootfs partition). I thought it was something unrelated I did, but maybe, some commands are escaping the chroot somehow?

The-going commented:
@150balbes Олег, как я могу воспроизвести это.

The-going commented:
I don't have that happening. If, when building an image for x86_64, the build system wants to install a dtb package that does not exist, then it is necessary to handle this exception. But in any case, the package cannot be installed on the machine where the assembly takes place.

The-going commented:
@rpardini Ricardo, I checked the build of images for x86_64, but only in the first edition of AR-1009. Nothing of the kind was noticed.

rpardini commented:
Thanks for the tests. I also can't reproduce it with current master, even with changes so the .dtbs are built again, but I worry that some old intermediary version might have somehow inadvertently caused it.

Especially because Armbian's build farm machines, Github Runners, indeed built most of the versions of the branch for all arches, including x86 (there's even _desktop_ x86 builds that I never thought about, very nice), so might also be affected.

So I advise @igorpecovnik and others who built the old AR-1009 branch or previous PR branch to be wary of reboot on those machines, maybe take a look at `tree /boot` before reboot and make sure the output of `update-grub` and the generated `/boot/grub/grub.cfg` makes sense, initrd is there, no dtbs mentioned, and rootfs is set via UUID not device name, before rebooting if possible. 

Hopefully before then, all will be stable enough for Armbian to build Armbian using Armbian, so good reason to reinstall builder ;-) Sorry @150balbes for the trouble, please report any other problems found.

150balbes commented:
> @150balbes Олег, как я могу воспроизвести это.

я заметил это несколько раз на сборочном хосте (система Ubuntu Hirsute в виртуалбоксе). началось это примерно с 1 декабря, при перезагрузке хоста, при старте, стало вылезать сообщение о не верной команде devicetree. полез смотреть grub.cfg и увидел, что в каталоге /boot (вся система стоит в одном разделе, может этот то же влияет) куча каталогов dtb-модель-плата и в конфиге grub прописана команда во многих местах /boot/devicetree. на этом хосте я ни когда не ставил ar1009 и не запускал ни чего из этой ветки, только обновление из ветки master (т.е. это прилетело точно из главной ветки после добавления в нее этих изменения). я не запускал на этом хосте ни каких сборок для x86 или EFI, только стандартные сборки образов для station и другие, которые я поддерживаю. после первого раза , когда вычистидл всё, через день обнаружил опять появление DTB каталогов (но уже меньше, несколько штук). за точность не поручусь, но появляется это , если запускаю сборку и происходит обновление и установка пакетов в хост систему самой сборочницей (изменились зависимости в конфигах сборочницы). скорее всего в этот момент, после установки пакетов на хост,  запускается штатная процедура на хосте обновления инитрд и конфига grub. возможно эта ошибка была в первой версии PR который слили с мастером, и сейчас изменились скрипты\конфигы и это не проявляется.

The-going commented:
@150balbes Олег, я использую `qemu+virt-manager`, поэтому следующий код у меня никогда не выполняется.
https://github.com/armbian/build/blob/e1a88959f8ad29083f7ab2878cb8cacaf823a8ae/compile.sh#L114-L124

Обходной путь:
сделай копию compile.sh с другим названием, закоментируй этот код, вычисти /etc/apt/sources.list (ни в одном файле /etc/apt/* не должно быть репозитория armbian) и обновляй виртуальную машину в ручную. 

Давно хотел спросить. Что в виртуалбоксе покажет команда:
`echo "$(systemd-detect-virt)"`

TRSx80 commented:
I just spent at least an hour studying the comments in this whole thread.

I really appreciate the care taken and that all the high level wizards are working together cooperatively on this PR.  And that documentation was "included" (heading over there next).

Cheers!

TRSx80 commented:
Well, for 2 hours I studied the code, which is _at best_ a cursory look at 35K lines of changes.  It's been merged already, so it's water over the dam at this point, but maybe next time we can split this up into some separate patch sets which would be more digestible for reviewers?  Perhaps along thematic lines (extensions framework, RPi stuff, x86 stuff, mvebu stuff, etc.)?
